### ISSUE TITLE
Add "Module" Support

### ISSUE THREAD

**Issue by augurs** [humhub/rest#33]
I am creating controller file under controllers folder in rest module and bind that controller in to rout in Event.php, but its not working it show 404 error. 

Can any one let me know how we implement new controller in HumHub Rest Module?

Thanks

**Comment by luke-** [comment]
I am not sure if this is currently possible. Probably we will have to introduce a new event for this.

**Comment by luke-** [comment]
Could be interesting to generate the documentation for a module: https://github.com/humhub/humhub-modules-rest/issues/9

**Comment by yurabakhtin** [comment]
@luke- 

> I am not sure if this is currently possible. Probably we will have to introduce a new event for this.

This is not implemented yet, this similar already was requested here https://github.com/humhub/humhub-modules-rest/issues/36#issuecomment-717908452.

As I understand we should replace all these lines https://github.com/humhub/humhub-modules-rest/blob/master/Events.php#L134-L138 into new event trigger:

```php
static::addWikiModuleRules();
static::addCalendarModuleRules();
static::addTasksModuleRules();
static::addCfilesModuleRules();
static::addMailModuleRules();
```

**Comment by luke-** [comment]
@yurabakhtin Yes we should trigger a new Event here and a module should then be able to add itself with endpoints.
But we should first implement this with one module on a prototype basis

**Comment by yurabakhtin** [comment]
@luke- What I have implemented in 2 PRs https://github.com/humhub/humhub-modules-rest/pull/55 and https://github.com/humhub/humhub-modules-mail/pull/217:

- Move all controller files of the module, e.g. from the folder `/humhub-modules-rest/controllers/mail` into the `/humhub-modules-mail/controllers/rest` (+ of course changes the controllers from `rest` namespace to the module namespace)
- Move all url rules of the module into the same event `Application::EVENT_BEFORE_REQUEST`, keep same pattern like `api/v1/mail` but change to new route, e.g. from `rest/mail/message/index` to new location `mail/rest/message/index`, so api endpoints will have same URLs after this upgrade. The rules should be appended only when the REST API module is enabled `if (Yii::$app->getModule('rest')) {`.


https://github.com/humhub/humhub-modules-mail/pull/217/files/2d326faa22875d258cfb30dce2c4939b2dab8576#r583499828 :

> Just as a thought: Could a separate new REST Module event be useful here? I am not sure if we would have any advantages here. The event could e.g. collect the routes as Event attribute.

Yes, my current solution doesn't require a new event trigger, and I find some modules already use the event `Application::EVENT_BEFORE_REQUEST` for other purposes. I need to investigate how to implement new separate event especially for the rest api url rules.

**Comment by yurabakhtin** [comment]
I think we should store the prefix `api/v1/` in some module const var like `const API_URL_PREFIX = 'api/v1/'`.
Also we can implement new method in the rest module in order to append url rules:

```php
Yii::$app->getModule('rest')->addRules([
    ['pattern' => 'mail', 'route' => 'mail/rest/message/index', 'verb' => 'GET'],
])
```

instead of current:

```php
Yii::$app->urlManager->addRules([
    ['pattern' => 'api/v1/mail', 'route' => 'mail/rest/message/index', 'verb' => 'GET'],
])
```

**Comment by yurabakhtin** [comment]
@luke- In commit https://github.com/humhub/humhub-modules-rest/pull/55/commits/3b987b2a9f95490caa0f04d092cae452dd8d0fd4 I have implemented new event `humhub\modules\rest\Module::EVENT_REST_API_ADD_RULES`.
Then I have used it in the mail module in commit https://github.com/humhub/humhub-modules-mail/pull/217/commits/6304b2a9f1cd23bc50c365a22dd556bc5ab81915, but please note it is impossible to use the event value as variable in config like `humhub\modules\rest\Module::EVENT_REST_API_ADD_RULES`, we can use only this as static string `restApiAddRules`, please see this line https://github.com/humhub/humhub-modules-mail/pull/217/commits/6304b2a9f1cd23bc50c365a22dd556bc5ab81915#diff-724326d4c9977d3f53ba0f053c7d857dc8b8ef4f2d61b2b377791b4ec560720dR18:

```php
['class' => 'humhub\modules\rest\Module', 'event' => 'restApiAddRules', 'callback' => ['humhub\modules\mail\Events', 'onRestApiAddRules']],
```

If this is ok and the event name `'restApiAddRules'` will not be changed in future we can use this way.

**Comment by luke-** [comment]
@yurabakhtin Okay, I suppose we'll have to do it that way.  But I wouldn't use a CamelCase here, and like Action IDs, use a format with dashes. `rest-api-add-endpoint`.

What do you think about an own event class like e.g. `RestAddEndpointEvent`. 
e.g.
https://github.com/humhub/humhub/blob/master/protected/humhub/modules/user/events/UserEvent.php#L25
There could be an array `urlRules` as array.

**Comment by yurabakhtin** [comment]
@luke- Additional improvements in commits https://github.com/humhub/humhub-modules-rest/pull/55/commits/eba89103f4e9087a4b9d4aac5c24f8fb674a4fb1 and https://github.com/humhub/humhub-modules-mail/pull/217/commits/796bee532c196a0120bbbc09765c0a91c35957be:

- new method `Yii::$app->getModule('rest')->addRules()`,
- new `const API_URL_PREFIX = 'api/v1/';` is used in the method,
- I did to use it as example in the `mail` module, and it can be used for core endpoints inside the `rest` module.

**Comment by luke-** [comment]
@yurabakhtin Ok, I also like it, then modules can directly use the `addRules` method.

**Comment by yurabakhtin** [comment]
@luke- 

> What do you think about an own event class like e.g. RestAddEndpointEvent.
e.g.
https://github.com/humhub/humhub/blob/master/protected/humhub/modules/user/events/UserEvent.php#L25
There could be an array `urlRules` as attribute.

Why should we pass an array `urlRules` as attribute from `rest` module into another module like `mail` if inside the event from `mail` module side we only need to **append** new rules by `Yii::$app->urlManager->addRules([...])`?
I am not sure we should allow to modify the core rules array inside other modules.
Please explain me what we should to do with the passed array inside the event method for example here https://github.com/humhub/humhub-modules-mail/blob/enh/33-rest-api-module/Events.php#L104 `public static function onRestApiAddRules()`, thanks.

**Comment by luke-** [comment]
Here is my idea:
```php 
    // Callback e.g. in Mail Modul e
    public static function onRestApiAddRules(RestAddEndpointEvent $event)
    {
            $event->addRules([...mail module rules...]);
            // or
           $event->urlManagerRules[] 
    }
```

But your approach is also good.
```php 
    public static function onRestApiAddRules()
    {
            /** @var \humhub\modules\rest\Module $module */
           $module = Yii::$app->getModule('rest');
           $restModule->addRule([.. my rules ...]);
    }
``` 

We can stay with your solution. Your choice.

**Comment by yurabakhtin** [comment]
@luke- Ah you wanted the event only to implement new method `RestAddEndpointEvent->addRules([])`.
If you really think we need that I can do that but don't see a reason.

**Comment by luke-** [comment]
@yurabakhtin Yes, exactly. Leave it as it is.

**Comment by yurabakhtin** [comment]
@luke- Commit https://github.com/humhub/humhub-modules-rest/pull/55/commits/e070611a5e123617148b072137eb10dd3adc9bc9 - I applied new method `Yii::$app->getModule('rest')->addRules([])` for all rules inside this module.
I think this task is completed.

**Comment by luke-** [comment]
@yurabakhtin Have you tried generating the Swagger documentation for a module?

Can you adapt the DEVELOPER.md documentation. It would be good if it says what a module developer needs to do to provide a REST api.

**Comment by yurabakhtin** [comment]
@luke-

> Have you tried generating the Swagger documentation for a module?

Ok, I have copied the `mail.yaml` and the generator files into the `mail` module - https://github.com/humhub/humhub-modules-mail/pull/217/commits/bf5ac33b44dc8f1af6f14636521f1295a535003f, but I please note there are many dependencies from another yaml files `common.yaml` and `user.yaml`, so I changed the lines like `$ref: '../../../rest/docs/swagger/common.yaml#/components/properties/totalProperty'`, however I couldn't find a way to make the path `../../../rest/docs/swagger/` variable, so it may be not easy to generate if the module is located in different folder as in my case it is located in the folder `humhub-modules-rest` instead of `rest`.

> Can you adapt the DEVELOPER.md documentation. It would be good if it says what a module developer needs to do to provide a REST api.

Commit https://github.com/humhub/humhub-modules-rest/pull/55/commits/82ee53000256bfc0788ab6138e415bab9440f4b0

**Comment by luke-** [comment]
@yurabakhtin We really need a solid "Module" solution for the REST documentation.

We probably have following options here:

- At the module level (e.g. mail module), create a build script that clones the REST module (Git) to provide the dependencies. Result would be a HTML file in the module repository with REST API Documentation. (Similar to the current approach.)
- On REST module level, create a build script that clones all modules (Git) with Swagger YAML files and builds the documentation for all modules centrally. e.g. Publish at (https://docs.humhub.org/rest)
- Find another format/system for the documentation. 
 -  ...?

Can you please do some fruther research here or do you already have a suggestion here?

**Comment by yurabakhtin** [comment]
@luke- 

> We really need a solid "Module" solution for the REST documentation.

Did you see https://github.com/humhub/humhub-modules-rest/commit/82ee53000256bfc0788ab6138e415bab9440f4b0 what I reported in my previous comment above? Is not enough for the REST documentation?

> We probably have following options here:

Sorry, but I don't understand why we need the additional scripts to clone modules. Could you tell me your idea with some examples.

Currently it already works just after installing the modules from back-office by admin.

**Comment by luke-** [comment]
@yurabakhtin Ahso sorry, by REST documentation I meant the automatically generated Swagger docs. 

It is about how we can generate the following page in the future:
https://www.humhub.com/en/marketplace/rest/#Manual

**Comment by yurabakhtin** [comment]
@luke- I have moved the swagger files of the `mail` module from the `rest` module into the `mail` module - commit https://github.com/humhub/humhub-modules-mail/commit/bf5ac33b44dc8f1af6f14636521f1295a535003f.
I.e. currently the mail module already has the folder for swagger files `yaml` and generated `html` file.

Here https://www.humhub.com/en/marketplace/rest/#Manual we have the text:

> ### Swagger / OpenAPI 3.0.0
>
> You can find Swagger documentation files in the `/docs/swagger` directory of this module.

For new version we can change to:

> You can find Swagger documentation files in the `/docs/swagger` directory of this module and of the external modules which support RESTful API endpoints.

PS: Postman folder was not copied into the `mail` module yet, it can be similarly but of course a new postman file should be created manually for each module.

**Comment by yurabakhtin** [comment]
@luke- 

> - REST API & Automatic Tests?

Do you mean to create some tests in the module "REST API"?

> - Is there a possibility to disable the REST API for some modules? e.g. as a use case I have a large installation and want to limit the REST API.

I have implemented this in commit https://github.com/humhub/humhub-modules-rest/pull/55/commits/e0d1ccff7fe89977cc37044900b62aaa8828e7b8, by default all modules with api endpoints are enabled, even if new module will be installed then it is also will be enabled by default when all previous modules are disabled:

<img width="856" alt="disable_api_modules" src="https://user-images.githubusercontent.com/6995524/110773739-1c6f4600-826e-11eb-85af-d70f53588a07.png">

From external modules with event `restApiAddRules` we should pass second param as module id like `Yii::$app->getModule('rest')->addRules([...], 'mail')` as I have done here https://github.com/humhub/humhub-modules-mail/pull/217/commits/eaf4f1d6c60e9220f742d0fbf1bca3a9607dade2.

---

> Otherwise can we proceed to implement this in the same style for all other modules?

Yes, I can do same for all other modules, firstly for not core modules https://github.com/humhub/humhub-modules-rest/blob/enh/33-move-external-module/Events.php#L140-L143: `wiki`, `calendar`, `tasks`, `cfiles`.
Then we can try to do same for core modules like `comment`, `post`, `user`, `space` and etc. but I'm not sure we need this.

**Comment by luke-** [comment]
@yurabakhtin About testing: It would be good if e.g. the Polls module could test its REST API endpoints somehow. Presumably, these are easily forgotten in the further development and therefore tests are particularly important here.

Cool, with the selectable modules. That's a really nice.

The core module thing is a good idea, we should probably think about that later.

**Comment by yurabakhtin** [comment]
@luke- I have started this in draft PR https://github.com/humhub/humhub/pull/4960 and also in commit https://github.com/humhub/humhub-modules-mail/pull/217/commits/21f3b18b3d920bfd26a6737cade1964f06ead38e, but it doesn't work properly yet, it works only when the table `users` is filled, but I cannot find how to do this automatically on run `codecept run api`. I tried to use `tests\codeception\_support\DynamicFixtureHelper`, `tests\codeception\_support\FixtureHelper` in `api.suite.yml` with other modules like they are used in other cases but it didn't help. Do you know a solution to use fixtures automatically for this new suite?

**Comment by luke-** [comment]
@yurabakhtin Unfortunately, I have no idea here either, but I'll take a look next week, maybe I'll find something here.

**Comment by luke-** [comment]
@yurabakhtin I just played around with it a bit. It worked for me to add the `DynamicFixtureHelper`. But after that you have to run `codecept build` again. did you do that?

**Comment by yurabakhtin** [comment]
@luke- Yes, I run `codecept build` after each changing config `yml` files as it is described there, but when I added `DynamicFixtureHelper` in config I received an unknown error like this:

<img width="1336" alt="test_api_with_user_fixtures" src="https://user-images.githubusercontent.com/6995524/111128052-8bf27780-8585-11eb-8b47-92d2be77f8d3.png">

and I thought the problem in fixtures however they were filled in DB, and only now after detailed debugging I could find the problem was in the REST module that I fixed here https://github.com/humhub/humhub-modules-rest/commit/294f0871bc8e4f698ef6fdd4a1494f9f1a7d47ad, so after that fix I can continue to work on next tests.

**Comment by luke-** [comment]
@yurabakhtin Ok, that was an ugly bug. Let me know if I can review anything, but it looks very good so far!

**Comment by yurabakhtin** [comment]
@luke- I have created tests for all endpoints of the module `mail` - https://github.com/humhub/humhub-modules-mail/commit/d7d791b9bcd657d574731f951a002fac7fc18eb5

**Comment by luke-** [comment]
@yurabakhtin Looks nice .As an idea before we roll it out for the other modules, do you think we should share REST outputs like the pagination between the modules? e.g. ->seePagination....

**Comment by yurabakhtin** [comment]
@luke- 

> Looks nice .As an idea before we roll it out for the other modules, do you think we should share REST outputs like the pagination between the modules? e.g. ->seePagination....

Not clear what you mean but I think you tell about API tests like this:

```php
$I->seeSuccessResponseContainsJson([
    'total' => 3,
    'page' => 1,
    'pages' => 1,
    'links' => [
        'self' => '/api/v1/mail/3/entries?page=1&per-page=100',
        'first' => '/api/v1/mail/3/entries?page=1&per-page=100',
        'last' => '/api/v1/mail/3/entries?page=1&per-page=100',
    ],
    'results' => [
        ['id' => 4, 'content' => 'Third Message entry text 1.', 'user_id' => 1],
        ['id' => 5, 'content' => 'Third Message entry text 2.', 'user_id' => 2],
        ['id' => 6, 'content' => 'Third Message entry text 3.', 'user_id' => 3],
    ],
]);
```

If you think we should not display the pagination data in the API tests we can use it like this and it works successfully as well:

```php
$I->seeSuccessResponseContainsJson([
    'results' => [
        ['id' => 4, 'content' => 'Third Message entry text 1.', 'user_id' => 1],
        ['id' => 5, 'content' => 'Third Message entry text 2.', 'user_id' => 2],
        ['id' => 6, 'content' => 'Third Message entry text 3.', 'user_id' => 3],
    ],
]);
```

Let me know if I should remove this from API tests.

**Comment by luke-** [comment]
@yurabakhtin I was just thinking about if we should introduce a helper for standard rest module responses. 
like e.g.the pagination from here:
https://github.com/humhub/humhub-modules-rest/blob/master/components/BaseController.php#L177

```php
$I->seePaginatedJsonResponse([
        ['id' => 4, 'content' => 'Third Message entry text 1.', 'user_id' => 1],
        ['id' => 5, 'content' => 'Third Message entry text 2.', 'user_id' => 2],
        ['id' => 6, 'content' => 'Third Message entry text 3.', 'user_id' => 3],
], [$total = null], [$page = 1]);
```
(Method name not ideal yet.)

Presumably, when implementing the API tests in other modules, we'll see whether there are redundant code parts that could be optimized.

**Comment by luke-** [comment]
@yurabakhtin Another example might be

- https://github.com/humhub/humhub-modules-rest/blob/master/definitions/UserDefinitions.php#L32
- https://github.com/humhub/humhub-modules-mail/pull/217/files#diff-84949359a85304fc9e914344d87c729fd194980a171ea64f72cf1b33da284466R71-R77

```php
$I->seeUserDefinition($userId)
```

**Comment by yurabakhtin** [comment]
@luke- For pagination tests I have created new methods in commit https://github.com/humhub/humhub/pull/4960/commits/3dbd8e2dbb4f732b3e2d6f7e4a4585adf74e630a and applied them here https://github.com/humhub/humhub-modules-mail/pull/217/commits/29bfdb2f6efbadc0324420cb48762bdaae03a824.

For user data I will do in next commits.

**Comment by yurabakhtin** [comment]
@luke- For user definitions: https://github.com/humhub/humhub/pull/4960/commits/4da7dfd7b573779283f6a5d2e84d7302225f1b4a + https://github.com/humhub/humhub-modules-mail/pull/217/commits/1d9d3e4745b70dff47e05d24c80d8c073aff18a9.

**Comment by luke-** [comment]
@yurabakhtin Looks good. If there are more cases, we can add the ApiTester class.

**Comment by yurabakhtin** [comment]
@luke- It seems you meant "we can add **into** the ApiTester class" because the ApiTester class already exists.

I think we can start to apply the changes for other external modules as we did for Mail module here https://github.com/humhub/humhub-modules-mail/pull/217/files.

**Comment by luke-** [comment]
@yurabakhtin Yes, there is "into" missing. Looks good, then let's start with further modules.

**Comment by luke-** [comment]
@yurabakhtin Yes, some tests for this would be great!

**Comment by yurabakhtin** [comment]
@luke- API tests for wiki module - https://github.com/humhub/humhub-modules-wiki/pull/166/commits/0cf85c8e7e2dcc769f7c7de5ffe0627ee1bfbdea.
I have created new PR https://github.com/humhub/humhub/pull/4976 but please read there:

> **NOTE:** I think we should wait when all external module API endpoints will be moved to its modules from RESTful API module, because probably the `ApiTester` will be changed.

---

Please note during work on the tests I detected the bugs in the following files:
- `controllers/rest/WikiController.php` https://github.com/humhub/humhub-modules-wiki/pull/166/commits/0cf85c8e7e2dcc769f7c7de5ffe0627ee1bfbdea#diff-593d2b9c6c83e9e6e4db789813b280d463da8fcebb4cd88ee52174bbcb087907
- `models/WikiPageRevision.php` https://github.com/humhub/humhub-modules-wiki/pull/166/commits/0cf85c8e7e2dcc769f7c7de5ffe0627ee1bfbdea#diff-608e690e6ed126b84d30459efa3c36c3e01229766e64c7fbb469256166fea717
- `models/forms/ItemDrop.php` https://github.com/humhub/humhub-modules-wiki/pull/166/commits/0cf85c8e7e2dcc769f7c7de5ffe0627ee1bfbdea#diff-a76d8d5e6a4ece4fc9b6b3015679b5337fdc51a4d778113e8e974af6bf798b26

I think they should be merged before we merge the PR with api changes, let me know to do a separate PR for them.

**Comment by luke-** [comment]
@yurabakhtin So I understand you correctly that we should wait with the Core PR merge because there are probably additional changes here? That's ok for me.

I can't estimate right now how critical the bugs in the wiki module are, if you think they are critical I would also prefer an extra PR with module release.

**Comment by yurabakhtin** [comment]
@luke- 

> So I understand you correctly that we should wait with the Core PR merge because there are probably additional changes here?

Yes, correct.

> I can't estimate right now how critical the bugs in the wiki module are, if you think they are critical I would also prefer an extra PR with module release.

I have reviewed the bugs again:

1. It is related to API endpoints, so no need to merge into `master` right now.
2. Here `WikiPageRevision->page` returns object of `User` instead of expected `WikiPage`, I cannot find where it is used, probably on history page, but currently I don't see any bugs there, but this wrong code broke one API request as I remember there was error like `object User has no properly "content"`, probably in this place https://github.com/humhub/humhub-modules-wiki/compare/enh/33-rest-api-module#diff-8a3671fa80cc05d1f0db43eeda5c46f2a9e8ea261f7d69c153f2eccf9e36dee1R60 `$revision->page->content->canEdit()`.
3. Can be reproduced only on tests, because on real working we have a redirect after that action so transaction is reset automatically.

So probably only 2nd issue can be merged, however I cannot reproduce this on normal(not tests, not api) work of the module.

**Comment by luke-** [comment]
@yurabakhtin Ok thanks, so then I would leave the merge for now.

Otherwise everything is clear here with this issue or are there open points? 
The target would be to migrate all the rest API endpoints into the modules and cover them with tests.

**Comment by yurabakhtin** [comment]
@luke- 

> Otherwise everything is clear here with this issue or are there open points?

Yes all clear, I just need a time to do same changes for other modules like already was done for `mail` and `wiki`.

**Comment by yurabakhtin** [comment]
@luke- I have completed the moving endpoints of all external modules, so the following PRs can be merged after we will merge core changes from PR https://github.com/humhub/humhub/pull/4976 and change of the REST module from https://github.com/humhub/humhub-modules-rest/pull/55.

PRs of the external modules:

- wiki: https://github.com/humhub/humhub-modules-wiki/pull/166
- calendar: https://github.com/humhub/humhub-modules-calendar/pull/235
- tasks: https://github.com/humhub/humhub-modules-tasks/pull/141
- cfiles: https://github.com/humhub/humhub-modules-cfiles/pull/111
- mail: https://github.com/humhub/humhub-modules-mail/pull/217

**Comment by luke-** [comment]
@yurabakhtin Ok great, then we should do following steps:

1.) Finish & Merge Core Changes (for Tests)
2.) Merge and Release all Modules 
3.) Release new REST API Module

**Comment by luke-** [comment]
@yurabakhtin I started with the wiki module, but can't get the GitHub action tests to work. Do you have any idea here?

**Comment by yurabakhtin** [comment]
@luke- I think you should do step "3.) Release new REST API Module". Please try to merge the PR https://github.com/humhub/humhub-modules-rest/pull/55 into master and then we will see how it works, because I see you added here https://github.com/humhub/wiki/commit/a7e87276595cbb3f832d630a09005710e31d212f#diff-acf3e0c6c881e3230ee20ea7499891845048d94362e17e496f06dafa7757f110R123 the using of the REST module but it seems there the branch `master` is used by default and there is no new necessary changes.
I hope this will helps because locally I get same errors when use the `master` branch of the REST module.

**Comment by luke-** [comment]
@yurabakhtin Good point!

I've changed the branch here: https://github.com/humhub/wiki/pull/167/commits/c949213f2622b7ccd76862506dbc4cd74fa083c6

Now the API test are working, but there are other errors:
https://github.com/humhub/wiki/runs/2280451477?check_suite_focus=true#step:25:32

Can you please take a look?

**Comment by yurabakhtin** [comment]
@luke- I have fixed the errors in tests with new PR https://github.com/humhub/humhub/pull/5002.

There were two bugs:
- `$I->amOnSpace($guid)` - from wiki module side the `$guid` was passed as string with real value of space guid but the function was not developed to receive such value, so I fixed it here https://github.com/humhub/humhub/pull/5002/files#diff-cd839f03c6f3acfd1b40a8f70488fae4a8b12ece053f9c97b8c520202ab8d554R251-R252
- Second issue was related to my recent code with `$this->haveFixtures(['space' => SpaceFixture::class]);` - this code must be load only once, because on second call it clear all related record of the already initalized space, so all settings like permissions were reset and a test user has no an expected permission to a requested page.

**Comment by luke-** [comment]
@yurabakhtin Thank you, now the Wiki module tests are running green. I've merged the changes from the REST module into the `develop` Branch for now, and I'll now release all the modules one by one.

**Comment by SebastianSchoeps** [comment]
Hey guys, don't want to disturb the intimacy here :-). I am just trying to understand if/when there will be a REST API for the polls module or what needs to get done to create one. I didn't really get it from what is written here or somewhere else. I thought this is the best place to ask. Thx!

**Comment by luke-** [comment]
@SebastianSchoeps Unfortunately, the Poll module does not have a REST API yet. We are migrating the module REST APIs to the respective module repositories. Previously, these were maintained centrally here in the REST API module repo.

However, the various module PRs are quite good templates to see what development is required for a module REST API. 
e.g. https://github.com/humhub/wiki/pull/166/files

For the Polls module I have opened an issue. 
https://github.com/humhub/polls/issues/86 - We are of course happy if you want to contribute something here. I can not yet estimate when from the Humhub team will find time for it.

**Comment by SebastianSchoeps** [comment]
Thanks for the quick feedback! Haven't worked with Yii yet, so cannot promise too much.
